---
title: "Tanner Review of Maggie Sodders"
output: html_document
---
 
# Hi Maggie, Great job on  the assignment. My suggestions are designated with a number symbol (#).

## R Homework Code 
Step 1: Get Data Into R
```{r}
library(tidyverse)
SNPData <- read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546-Spring2021/main/assignments/UNIX_Assignment/snp_position.txt")
FangData <- read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546-Spring2021/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt")
```
	Step 2: Data Inspection
# You can also use >tail to ensure that nothing weird is added to the end of the file as well as dimensions to determine the size of the table.
tail(SNPData)
tail(FangData)
dim(SNPData)
dim(FangData)
```{r}
  ncol(SNPData)
  ncol(FangData)
  nrow(SNPData)
  nrow(FangData)
	typeof(SNPData)
	typeof(FangData)
	names(SNPData) 
	names(FangData)
```
Step 3: Data Processing 
*Separate Maize and Teosinte Data In Fang File and Head to check them*
#This was slick, you could search on just =="ZMM" and =="ZMP", but your approach allows you to be a lot more specific in the event three letters is not enough to distinguish the species.
```{r}
FangData_Maize <- FangData[which(FangData$Group=="ZMMIL" | FangData$Group =="ZMMLR" | FangData$Group == "ZMMMR"),]
	head(FangData_Maize)
		FangData_Teosinte <-FangData[which(FangData$Group=="ZMPBA" | FangData$Group =="ZMPIL" | FangData$Group == "ZMPJA"),]
	head(FangData_Teosinte)
```
*Prep SNP data*

#I liked that you preprocessed the SNP_Data.
```{r}
SNPData_SpecificC <- SNPData[,c(1,3,4)]
```
*Transpose Data*
```{r}
MaizeData_transposed <- as.data.frame(t(FangData_Maize))
TeosinteData_transposed <- as.data.frame(t(FangData_Teosinte))
```
*Trim Rows and Name Column*
```{r}
	colnames(MaizeData_transposed) <- as.character(unlist(MaizeData_transposed[1,])) 
	colnames(TeosinteData_transposed) <- as.character(unlist(TeosinteData_transposed[1,]))
	MaizeData_trim <- MaizeData_transposed[-c(1:3),]
	TeosinteData_trim <- TeosinteData_transposed[-c(1:3),]
```
*Join SNP and Fang Data*

# I like your approach here to merge the files. For me, however, I like having  the SNP_ID's out of the row name columns in it's own file thus allowing me to work with it in the future if I need it. Below is the code for that:
# X = rownames_to_column(TeosinteData_trim, 'SNP_ID')
# I would then merge the two file on "SNP_ID'.
# Join_SNP_Geno= right_join(SNPData_SpecificC,X, by ='SNP_ID')
```{r}
MaizeData_joined <- merge(x = SNPData_SpecificC, y = MaizeData_trim, by.x = "SNP_ID", by.y ="row.names", all.y = TRUE)
TeosinteData_joined <- merge(x = SNPData_SpecificC, y = TeosinteData_trim, by.x = "SNP_ID", by.y ="row.names", all.y = TRUE)
```
*Make SNP Increasing*
# Good-Very Slick approach
```{r}
	MaizeData_joined <- MaizeData_joined[order(MaizeData_joined$Position),] 
	TeosinteData_joined <- TeosinteData_joined[order(TeosinteData_joined$Position),]
```
*Put Each increase chromosome in its own .csv File*

# I like how efficient this script is! I am curious if the first row is needed, however. It appears that Sample ID from the FangData file is left over. 

# Wherever the origin of the Sample_IDs is for data processing. You can remove similar to this:
#Headerless_Transposed_ZMM_Grouped_Fang_geno= Transposed_ZMM_Grouped_Fang_geno[-c(1,2,3),]
# I will add however, it is a nice addition when looking at the Excel file so it may just be a preference for future analysis. 
```{r}
	for (a in 1:10){
  		MaizeData_Chromosome_Ordered <- MaizeData_joined[MaizeData_joined$Chromosome == a, ]
  		write.csv(MaizeData_Chromosome_Ordered, file= paste("Maize_chromosome", a, ".csv", sep=""), row.names = F)
		}
	for (a in 1:10){
  		TeosinteData_Chromosome_Ordered <- TeosinteData_joined[TeosinteData_joined$Chromosome == a, ]
  		write.csv(TeosinteData_Chromosome_Ordered, file= paste("Teosinte_chromosome", a, ".csv", sep=""), row.names = F)
		}
```
*Make SNP Decreasing*
#Good Stuff
```{r}
MaizeData_joined_decrease <- MaizeData_joined[order(MaizeData_joined$Position, decreasing=T),]

TeosinteData_joined_decrease <- TeosinteData_joined[order(TeosinteData_joined$Position, decreasing=T),]
```
*Put Each decrease chromosome in its own .csv File*
#Good Stuff
```{r}
for (a in 1:10){
 		MaizeData_DecreaseOrdered<-MaizeData_joined_decrease[MaizeData_joined_decrease$Chromosome == a, ]
  		write.csv(MaizeData_joined_decrease, file= paste("Maize_chromosome", a, "d.csv", sep=""), row.names = F)
		}
	for (a in 1:10){
 		TeosinteData_DecreaseOrdered<-TeosinteData_joined_decrease[TeosinteData_joined_decrease$Chromosome == a, ]
  		write.csv(TeosinteData_joined_decrease, file= paste("Teosinte_chromosome", a, "d.csv", sep=""), row.names = F)
		}
```
Step 4: Data Visualization 
*Package Installation/Setup*
```{r}
library(ggplot2)
library(tidyverse)
library(reshape2)
library(dplyr)
library(plyr)

FangData_transposed2 <- as.data.frame(t(FangData))
colnames(FangData_transposed2) <- as.character(unlist(FangData_transposed2[1,]))
FangData_SNPData_joined <- merge(x = SNPData_SpecificC, y = FangData_transposed2, by.x = "SNP_ID", by.y ="row.names", all.y = TRUE)
```
SNP/Chromosome

#Don't forget Distribution of SNPS 
#ggplot(data = Cleaned_Join_SNP_Geno) + geom_point(mapping = aes(x=Position, y=Chromosome))
```{r}
	ggplot(FangData_SNPData_joined, aes((Chromosome))) + geom_bar() + ggtitle("SNPs Per Chromosome") + labs(x="Chromosome",y="SNP Number")
```
Missing data and amount of Heterozygosity
```{r}
headers <- colnames(FangData)[-c(1:3)]
FangData_melted <- melt(FangData, measure.vars = headers)
FangData_melted[ FangData_melted == "?/?" ] = NA
FangData_melted$isHomozygous <- (FangData_melted$value=="A/A" | FangData_melted$value=="C/C" | FangData_melted$value=="G/G" | FangData_melted$value=="T/T")
```
*Ordered*
```{r}
FangData_melted_bysample <- FangData_melted[order(FangData_melted$Sample_ID),]
```
*Counted*
```{r}
FangData_bysample_count <- ddply(FangData_melted_bysample, c("Sample_ID"), summarise, counting_homozygous=sum(isHomozygous, na.rm=TRUE), counting_heterozygous=sum(!isHomozygous, na.rm=TRUE), isNA=sum(is.na(isHomozygous)))

FangData_bysample_count_melted <- melt(FangData_bysample_count, measure.vars = c("counting_homozygous", "counting_heterozygous", "isNA"))
```
*Order Better for Plotting*
```{r}
FangData_grouped <- FangData_melted[order(FangData_melted$Group),]
FangData_grouped_counted <- ddply(FangData_grouped, c("Group"), summarise, counting_homozygous=sum(isHomozygous, na.rm=TRUE), counting_heterozygous=sum(!isHomozygous, na.rm=TRUE), isNA=sum(is.na(isHomozygous)))
FangData_grouped_counted_melted <- melt(FangData_grouped_counted, measure.vars = c("counting_homozygous", "counting_heterozygous", "isNA"))
```
*Plotting2*
```{r}
ggplot(FangData_grouped_counted_melted,aes(x = Group, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack")
```
Your own visualization- Plot of how many of each group of SNP is contained in FangData
```{r}
ggplot(data = FangData_grouped) + geom_density(mapping = aes(x=value), fill="blue")
```

	# Good Work Maggie! You seem to be a very proficient user of R, your scripts are simple yet powerful. 
	


